### ISSUE TITLE
Unable to view localhost when "concurrentFeatures" is enabled. Requests are left to time out.

### ISSUE THREAD

**Issue by mattjis** [vercel/next.js#30549]
### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

15.4.0

### What browser are you using?

Safari and Chrome

### What operating system are you using?

macOS Big Sur 11.6

### How are you deploying your application?

vercel dev

### Describe the Bug

I'm trying to enable server components on a relatively fresh project of mine.

I ran `npm install react@alpha react-dom@alpha --legacy-peer-deps` and then added the following to my `next.config.js` file:
```
experimental: {
  concurrentFeatures: true,
  serverComponents: true
}
```

I was then getting the following error:
```
error - ./node_modules/next/dist/build/webpack/loaders/next-middleware-ssr-loader/index.js?page=-web%2Fabout&absolutePagePath=%2FUsers%2Fmattijs%2FDocuments%2FGitHub.nosync%2Forder-magazine%2Fpages%2Fabout.js&isServerComponent=false&buildId=development&basePath=&assetPrefix=!
Module not found: Can't resolve '/Users/mattijs/Documents/GitHub.nosync/order-magazine/pages/_document' in '/Users/mattijs/Documents/GitHub.nosync/order-magazine'
error - Error: ENOENT: no such file or directory, open '/Users/mattijs/Documents/GitHub.nosync/order-magazine/.next/server/middleware-ssr-runtime.js'
```

So I created a `_document.js` file with the following:
```
import { Html, Head, Main, NextScript } from "next/document"

export default function Document() {
	return (
		<Html>
			<Head />
			<body>
				<Main />
				<NextScript />
			</body>
		</Html>
	)
}
```

And now I'm getting no alarming errors only the following:
```
error - ReferenceError: window is not defined
Error! Failed to complete request to /about: Error: aborted
Error! Failed to complete request to /about: Error: socket hang up
```

Basically, if I try visiting `http://localhost:3000` the page will never load and eventually time out.

### Expected Behavior

I expect to be able to view the preview of my site, and play around with the new server components.

### To Reproduce

See "describe the bug" section.

**Comment by huozhi** [comment]
hi @mattjis are you sure `_document.js` is under `pages/` folder? I cannot repro this locally

**Comment by mattjis** [comment]
@huozhi I'm experiencing two issues really. 

The first is when I don't have a `_document.js` file of my own under the `/pages/` directory. I've never wanted this file, but now without it I get an error.

The second error (after resolving the missing `_document.js` file issue) is that I cannot view my `localhost`. I don't have much information on the reasoning behind this. The only thing I know is that it breaks when I add the `concurrentFeatures: true` feature to my `next.config.js` file.

**Comment by mattjis** [comment]
@huozhi If I keep refreshing multiple times I can get a memory leak to show up. 

Each refresh triggers `error - ReferenceError: window is not defined`

```
error - ReferenceError: window is not defined
error - ReferenceError: window is not defined
(node:7979) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 server-web/_error listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
```

**Comment by huozhi** [comment]
For the 1st issue, in the early release of next 12 we required a custom _document component but now we've provided a fallback for it. You can try the latest next@12.0.2 to see if the error is still preserved?

For the 2nd issue can you provide a minimal reproduction that we can debug with to trace the issue? Thanks

**Comment by mattjis** [comment]
@huozhi so I've figured out which import is causing this hanging behaviour.

It is from `import sanityClient from "@sanity/client"` in my` index.js` page.

**Comment by mattjis** [comment]
@huozhi here's a repo (which is based on the next.js starter) that imports `@sanity/client` to demonstrate the issue. 

https://github.com/mattjis/nextjs-concurrent-issue

**Comment by huozhi** [comment]
With `concurrentFeatures` enabled, pages are been deployed to edge functions which doesn't have browser client variables like `window`. 
Since you set `serverComponents` to `true` guess you wanna try RSC. One way is: you can change `pages/index.js` to `pages/index.server.js` and import `sanity` lib in a client component (`xxx.client.js`) and let it be imported by `index.server.js`. But notice that in this case CSS modules and `next/head` are not supported.

**Comment by mattjis** [comment]
@huozhi I've only ever used the `@sanity/client` server side. e.g in `getStaticProps`. I don't want to have to make these requests from the client side.

**Comment by shuding** [comment]
One workaround is, instead of doing

```js
import sanityClient from "@sanity/client"
```

You can try

```js
if (typeof window !== 'undefined') {
  require('@sanity/client')
}
```

But we still need to investigate why it wasn't an issue before — `window` is undefined when we do normal pre-rendering.

**Comment by mattjis** [comment]
@shuding not really a workaround as it's just disabling sanity. 
I need to be able to use `sanityClient` in the component to get data for it.

Sorry if my example repository is limited. Sanity requires API keys etc to give a real demo which makes it difficult.

**Comment by mattjis** [comment]
<img width="1175" alt="Screen Shot 2021-11-05 at 12 41 00 am" src="https://user-images.githubusercontent.com/21289982/140448303-deff2a90-b397-4b8e-bea7-90af276ef1d4.png">

**Comment by huozhi** [comment]
Looks like the browser entry of the library is resolved and run in the custom web runtime, which has some side effects to directly access window that cause this error. https://github.com/sanity-io/sanity/blob/a768eaf2611b4124e0ed6525bc82b0aa7e0ce00b/packages/%40sanity/eventsource/browser.js

**Comment by mattjis** [comment]
@huozhi does that mean the solution is within Sanity? And not Next?

**Comment by diegohaz** [comment]
I'm seeing this error with several packages that ship different browser and node bundles (using the `browser` field on package.json). It looks like Next.js is always getting the browser bundles when running on the server with `concurrentFeatures` enabled.

**Comment by huozhi** [comment]
For the libraries that are designed for browser and node runtimes separately while running between client and server side, like:

```
client: browser bundle
server: node bundle
```

They're not able to be adopted on this experimental feature so far since current streaming is running inside a custom web runtime (browser like) but without things that are special to web (like DOM, browser specific APIs). In short term, unluckily you cannot use them such as sanity.

We're working on letting users play with more node runtime libraries now, then we're able to try `concurrentFeatures` with sanity in the near feature.

**Comment by diegohaz** [comment]
@huozhi Thanks for the explanation! I just saw this error when trying to use `react-fetch@experimental`, which I suppose will be a common library for data fetching on the server using React Suspense. But I remember seeing this error even with `react-dom` (probably because another browser library imported it).

**Comment by huozhi** [comment]
@diegohaz I patched a fix for `react-fetch` yesterday now you can use latest experimental (`>= 0.0.0-experimental-f320ef88f-20211116`) version  of `react`, `react-dom` and `react-fetch` to play with next.js with `concurrentFeatures`.

```js
import { Suspense } from "react";
import { fetch as useFetch } from "react-fetch";

function Index() {
  const data = useFetch('/api/list').json()
  return <div>{data.list.join(',')}</div>
}

export default function IndexPage() {
  return <Suspense fallback={`loading`}><Index /></Suspense>
}
```

**Comment by huozhi** [comment]
We have a new option in experimental which is `runtime: 'nodejs' | 'edge'` which could let you run in nodejs runtime. Should be able to resolve the 3rd party library issue on edge runtime.

@mattjis Can you try to change your `concurrentFeatures: true` flag to `runtime: 'nodejs'` to try again? Thanks
